pkg/ccl/backupccl/restore_job.go

```diff
@@ -448,6 +463,40 @@ func restore(
 		tasks = append(tasks, generativeCheckpointLoop)
 	}
 
+	procCompleteCh := make(chan struct{})
+	// countCompletedProcLoop is responsible for counting the number of completed
+	// processors. The goroutine returns a retryable error such that the job can
+	// replan if there are too many lagging nodes.
+	countCompletedProcLoop := func(ctx context.Context) error {
+		var numProcCompleted int
+		tick := time.NewTicker(replanFrequency.Get(&execCtx.ExecCfg().Settings.SV))
+		defer tick.Stop()
+		done := ctx.Done()
+
+		for {
+			select {
+			case <-done:
+				return ctx.Err()
+			case _, ok := <-procCompleteCh:
+				if !ok {
+					return nil
+				}
+				numProcCompleted++
+			case <-tick.C:
+				// Replan the restore job if only a single processor has completed due to
+				// unbalanced workload given by the current DistSQL plan. If a processor
+				// completes the same time while receiving a tick, exclude the processor
+				// from the current interval, i.e., return retryable error if
+				// 1 <= numProcCompleted <= 2.
+				if numProcCompleted == 1 || numProcCompleted == 2 {
+					return errors.Mark(laggingRestoreProcErr, retryableRestoreProcError)
```

Comment:
I'm not sure this is really what we want. Say we have 10 nodes and 3 of them complete all at the same time.  In this case, we'll never replan based on lagging nodes even if 20 minutes later we are down to 1 node doing all of the work.

Perhaps we could structure this a bit differently. What if instead of using a ticker, we use a timer and every time we get a completion message, we reset the timer. If we ever read off the timer channel, that means it has been >= 10 minutes since the last processor stop.

For big clusters this might not catch situations where we have a lot of nodes with 1 node finishing every few minutes that would still benefit from replanning, but, it seems like it might be an easier place to start.

Reply:
This makes sense to me, I updated the code to use timer instead.